### PR TITLE
Do not propagate access rights to team members

### LINKED ISSUE #485: No access right propagation to team members
## Motivations

Teams are meant to be used by user to participate "as a team" instead of working alone. If permissions given to teams are propagated to users, these users will have the possibility to do what admin wanted the teams (so team users as a team) to be able to do.
For instance, the `can_view:content` granted to a team when entering a contest could be used by the individual users to access the contest tasks out of the team context.

So we want not to propagate permissions from teams to its members. 

## Options
1. Either not consider team-user relationships in the groups_ancestors table, so not including users in a team in the groups_ancestors table
2. ~~Or not consider team members as descendants of team ancestors, so keep team-users relations and team-ancestors in groups_ancestors, but not connect both.~~
3. ~~Or mark relationship to user via a team with a special flag so that we can use.~~

## Impact on services
* in "enter": the computation of additional time for a user will take into consideration the additional time given to its teams (and ancestors) --> expected behavior, nothing to change
* computation of additional time: expected behavior: additional time given to team or its ancestors is not to be given to the user alone 
* in enter/qualification_state: the computing of how many members can enter is based of perm of individual users  --> expected behavior, nothing to change
* groupChildrenView: the user count has to take into account the users in team
* getGroupProgress: simplify the computation of endMembers (which could be called "participants") by just taking the descendants which are users or teams (as users in team will not be descendants anyway).
* getUserProgress: Should be ok as the service was already handling separately users alone and users in teams. 
* getGroup: a user sees everybody is descendant from -> ok, ignore the ancestors of its teams that he is not descendants of by other paths
* getUserDescendants:  to return all descendant users, (on option 1 and 2) return the descendant users and the children of the descendant teams
* in `getActiveContestItemIDForUser`, the groups_ancestors seems to be used to match either the user or its team (? to be checked) Adapt it.
* getRecentActivity should not use `WhereUsersAreDescendantsOfGroup` but a condition on the group of the attempt which should be a descendant

## Other changes:
Update `groups_ancestors` table comment: "All ancestor relationships for groups, given that a group is its own ancestor and team ancestors are not propagated to their members . It is a cache table that can be recomputed based on the content of groups_groups."

### PR DESCRIPTION
**Pull request by zenovich**
Fixes #485

### CODE REVIEW

File: app/database/item_store.go [hunk 1]
```diff
@@ -292,15 +292,14 @@ func (s *ItemStore) getActiveContestItemIDForUser(user *User) *int64 {
...
 		Joins(`
 			JOIN groups_groups_active
 				ON groups_groups_active.parent_group_id = items.contest_participants_group_id AND
-					groups_groups_active.child_group_id = groups_ancestors_active.child_group_id`).
+					(groups_groups_active.child_group_id = user_link.parent_group_id OR groups_groups_active.child_group_id = ?)`,
+			user.GroupID).
 		Joins(`JOIN results AS contest_participations ON contest_participations.item_id = items.id AND
-			contest_participations.participant_id = groups_ancestors_active.ancestor_group_id AND
+			contest_participations.participant_id = groups_groups_active.child_group_id AND
 			contest_participations.started_at IS NOT NULL`).
```
[the remark this answers is not in the document]
Comment (zenovich) [reply app/database/item_store.go]: fixed

File: app/api/answers/submit.go [hunk 2]
```diff
@@ -27,7 +27,8 @@ import (
 //
 //   * The user should have submission rights on `task_token.idItemLocal`.
 //
-//   * The attempt should allow submission (`allows_submissions_until` should be a time in the future).
+//   * The attempt should allow submission (`results.started_at != NULL and
```
Comment (smadbe) [inline app/api/answers/submit.go:30]: The `started_at` is not really about allowing submissions. One should not arrive here (having a token) without the result being started, that's all.

Comment (zenovich) [reply app/api/answers/submit.go]: fixed

File: app/api/contests/get_qualification_state.go [hunk 3]
```diff
@@ -261,7 +268,7 @@ func (srv *Service) getQualificatonInfo(groupID, itemID int64, user *database.Us
 			Order("groups_groups_active.child_group_id").
 			Select(`
 				users.first_name, users.last_name, users.group_id AS group_id, users.login,
-				IFNULL(MAX(permissions_granted.can_enter_from <= NOW() AND NOW() < permissions_granted.can_enter_until), 0) AND
+				(MAX(team.can_enter) OR IFNULL(MAX(permissions_granted.can_enter_from <= NOW() AND NOW() < permissions_granted.can_enter_until), 0)) AND
```
[the remark this answers is not in the document]
Comment (zenovich) [reply app/api/contests/get_qualification_state.go:271]: Can we keep this "as is" for now?

[the remark this answers is not in the document]
Comment (smadbe) [reply app/api/contests/get_qualification_state.go:271]: Ok if you want, but I actually prefer the previous version (will be easier to adapt for #502) than the current more complex one.

**Pull request merged**